util/configowner.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/controllers/external"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ConfigOwner provides a data interface for different config owner types.
+type ConfigOwner struct {
+	*unstructured.Unstructured
+}
+
+// InfrastructureReady extracts infrastructure status from the config owner.
+func (co ConfigOwner) InfrastructureReady() bool {
+	infrastructureReady, _, err := unstructured.NestedBool(co.Object, "status", "infrastructureReady")
+	if err != nil {
+		return false
+	}
+	return infrastructureReady
+}
+
+// ClusterName extracts spec.clusterName from the config owner.
+func (co ConfigOwner) ClusterName() string {
+	clusterName, _, err := unstructured.NestedString(co.Object, "spec", "clusterName")
+	if err != nil {
+		return ""
+	}
+	return clusterName
+}
+
+// DataSecretName extracts spec.bootstrap.dataSecretName from the config owner.
+func (co ConfigOwner) DataSecretName() *string {
+	dataSecretName, exist, err := unstructured.NestedString(co.Object, "spec", "bootstrap", "dataSecretName")
+	if err != nil || !exist {
+		return nil
+	}
+	return &dataSecretName
+}
+
+// IsControlPlane checks if an unstructured object is a control plane node.
+func (co ConfigOwner) IsControlPlane() bool {
```

Comment:
Should this be more like `IsControlPlaneMachine` instead? We'd include the current check for labels, and the check that the owner's `Kind` is a Machine.